torch/csrc/jit/constants.cpp

```diff
@@ -20,8 +20,10 @@ Value* insertConstant(
       n->destroy();
       return g.insertNode(g.createUndefined())->output();
     }
-    if (ref.is_variable()) {
-      ref = autograd::Variable(ref).data();
+    if (!ref.is_variable()) {
+      ref = autograd::make_variable(ref, /*requires_grad=*/false);
+    } else {
+      ref.set_requires_grad(false);
```

Comment:
no one should be passing a requires_grad tensor here, so this should be an assert as well. I don't want to silently fix places like this, because it covers up larger logic errors in how constants are created. If it requires_grad it is also likely someone else might mutate it, which would be bad as well.

Reply:
Actually, I think this test is bogus. We should not be silently turning parameters into constants. I think what should happen here is a nice user-facing error that says that code is attempting to use a gradient recording tensor as a constant in a trace, and that the user likely wanted to trace the module rather than a function calling the module. I think it is arguable that we should not be implicitly turning anything into constants, but that is a larger change to consider.

Reply:
@zdevito Got it and I can remove the `test_call_python_mod_from_tracing_fn` test.
The other two failing tests look like the following:
```python
def test_call_script_mod_from_tracing_fn(self):
    class ScriptMod(torch.jit.ScriptModule):
        def __init__(self):
            super(ScriptMod, self).__init__()
            self.param = torch.nn.Parameter(torch.rand(4, 3))

        @torch.jit.script_method
        def forward(self, x):
            return torch.mm(x, self.param)

    sm = ScriptMod()

    @_trace(torch.rand(3, 4))
    def traced_fn(x):
        return sm(x) + 1.0

    self.assertExpected(canonical(traced_fn.graph))
```
```python
def test_call_traced_mod_from_tracing_fn(self):
    class TracedModule(torch.nn.Module):
        def __init__(self):
            super(TracedModule, self).__init__()
            self.param = torch.nn.Parameter(torch.rand(4, 3))

        def forward(self, x):
            return torch.mm(x, self.param)

    tm = torch.jit.trace(TracedModule(), torch.rand(3, 4))

    @_trace(torch.rand(3, 4))
    def traced_fn(x):
        return tm(x) + 1.0

    # Note: the parameter self.param from the Python module is inlined
    # into the graph
    self.assertExpected(canonical(traced_fn.graph))
```
These two tests have the same issue in that `self.param = torch.nn.Parameter(torch.rand(4, 3))` has `requires_grad=True` by default (and if I manually set `self.param` to `requires_grad=False`, the test will pass). Are the use cases in these two tests valid / should we also remove them?

Reply:
We should keep these tests because they are checking that some interactions between tracing and script work. However, it is fine to explicitly set requires_grad=False to make them pass. That is not what they were testing before.

Reply:
That might be true of the first test as well. We can just turn of the requires_grad to make them all pass.